tools/oeedger8r/Emitter.ml

```diff
@@ -967,274 +472,861 @@ let warn_size_and_count_params (fd : func_decl) =
+                && (is_in_ptr p || is_inout_ptr p) )
+              fd.plist)
+       in
+       if params <> [] then String.concat "\n" params
+       else "    /* There were no in nor in-out string parameters. */")
+    ; ""
+    ; "    /* lfence after checks. */"
+    ; "    oe_lfence();"
     ; ""
-    ; "/**** OCALL function wrappers. ****/"
-    ; String.concat "\n" oe_gen_enclave_ocall_wrappers
+    ; (* Call the enclave function *)
+      "    " ^ String.concat "\n    " (oe_gen_call_user_function fd)
     ; ""
-    ; "OE_EXTERNC_END"
+    ; (* Mark call as success *)
+      "    /* Success. */"
+    ; "    _result = OE_OK;"
+    ; "    *output_bytes_written = output_buffer_offset;"
+    ; ""
+    ; "done:"
+    ; "    if (pargs_out && output_buffer_size >= sizeof(*pargs_out))"
+    ; "        pargs_out->_result = _result;"
+    ; "}"
     ; "" ]
   in
-  let ecalls_fname = ec.file_shortnm ^ "_t.c" in
-  let os = open_file ecalls_fname ep.trusted_dir in
-  fprintf os "%s" (String.concat "\n" content) ;
-  close_out os
-
-let gen_u_h (ec : enclave_content) (ep : edger8r_params) =
-  let oe_gen_tfunc_wrapper_prototypes (tfs : trusted_func list) =
-    if tfs <> [] then
-      List.map
-        (fun f -> sprintf "%s;" (oe_gen_wrapper_prototype f.tf_fdecl true))
-        tfs
-    else ["/* There were no ecalls. */"]
-  in
-  let oe_gen_ufunc_prototypes (ufs : untrusted_func list) =
-    if ufs <> [] then
-      List.map (fun f -> sprintf "%s;" (oe_gen_prototype f.uf_fdecl)) ufs
-    else ["/* There were no ocalls. */"]
+  let gen_fill_marshal_struct (fd : func_decl) =
+    (* Generate assignment argument to corresponding field in args. This
+       is necessary for all arguments, not just copy-as-value, because
+       they are used directly by later marshalling code. *)
+    let rec gen_assignment prefix count (ptype, decl) =
+      let id = decl.identifier in
+      let arg = prefix ^ id in
+      [ gen_c_for count
+      ; [ sprintf "_args.%s = %s%s;" arg
+            (get_cast_to_mem_expr (ptype, decl) true)
+            arg ]
+      ; (* for string parameter fill the len field *)
+        ( if is_str_ptr ptype then
+          [sprintf "_args.%s_len = (%s) ? (strlen(%s) + 1) : 0;" arg id id]
+        else if is_wstr_ptr ptype then
+          [sprintf "_args.%s_len = (%s) ? (wcslen(%s) + 1) : 0;" arg id id]
+        else [] )
+      ; (let param_count =
+           oe_get_param_count (ptype, decl, "_args." ^ prefix)
+         in
+         flatten_map
+           (gen_assignment (arg ^ gen_c_deref param_count) param_count)
+           (get_deepcopy (get_param_atype ptype))) ]
+      |> List.flatten
+    in
+    flatten_map (gen_assignment "" "1") fd.plist
   in
-  let guard = sprintf "EDGER8R_%s_U_H" (String.uppercase ec.file_shortnm) in
-  let content =
-    [ sprintf "#ifndef %s" guard
-    ; sprintf "#define %s" guard
+  (* Generate host ECALL wrapper function. *)
+  let oe_gen_host_ecall_wrapper (name : string) (tf : trusted_func) =
```

Comment:
The `name` parameter is ripe for refactoring after the nesting of these functions.